mllib-tests/v1p5/src/main/scala/mllib/perf/MLAlgorithmTests.scala

```diff
@@ -490,14 +502,34 @@ abstract class DecisionTreeTests(sc: SparkContext)
 
   protected var labelType = -1
 
-  def validate(model: Either[RandomForestModel, GradientBoostedTreesModel], rdd: RDD[LabeledPoint])
-    : Double = {
+  def validate(model: DecisionTreeModel, rdd: RDD[LabeledPoint]): Double = {
+    val sqlContext = new SQLContext(rdd.sparkContext)
     val numExamples = rdd.count()
-    val predictions: RDD[(Double, Double)] = rdd.map { example =>
-      model match {
-        case Left(rfModel) => (rfModel.predict(example.features), example.label)
-        case Right(gbtModel) => (gbtModel.predict(example.features), example.label)
-      }
+    val predictions: RDD[(Double, Double)] = model match {
+      case MLlibRFModel(rfModel) =>
+        rdd.map(example => (rfModel.predict(example.features), example.label))
+      case MLlibGBTModel(gbtModel) =>
+        rdd.map(example => (gbtModel.predict(example.features), example.label))
+      case MLRFRegressionModel(rfModel) =>
```

Comment:
OK; actually matching whether the case class contains a `PredictionModel` would require using structural typing, which is implemented by reflections and hence would introduce significant overhead for the ML tests. Instead of implementing this as a pattern match, I'm going to avoid the reflections overhead by refactoring out a generic ML model `makePredictions` method.